### ISSUE TITLE
Supporting "full" branch definitions for build triggers.

### ISSUE THREAD

**Issue by auriusbendikas** [TrimbleSolutionsCorporation/TeamcityTriggerHook#24]
Hi,

If using the BitBucket branching model with default branch prefixes, the current version of plugin only supports build triggers when TeamCity VCS branch definitions is one of the following or combination of the following:
```
+:refs/heads/(master)
+:refs/heads/feature/(*)
+:refs/heads/bugfix/(*)
+:refs/heads/hotfix/(*)
+:refs/heads/release/(*)
```
However our project uses different TeamCity branch definitions (below). And those definitions are the only ones supported by [Sonar for Bitbucket Server](https://marketplace.atlassian.com/plugins/ch.mibex.stash.sonar4stash/server/overview) plugin. And due to the nature of our build server configuration currently using "VCS" trigger does not work in a satisfactory fashion.
```
+:refs/heads/(master)
+:refs/heads/(feature/*)
+:refs/heads/(bugfix/*)
+:refs/heads/(hotfix/*)
+:refs/heads/(release/*)
```
Support for above mentioned VCS branch definitions is required while using "build" trigger.

**Comment by auriusbendikas** [comment]
I have created a quick a and dirty fix on my forked branch [dirty-quick-fix](/auriusbendikas/TeamcityTriggerHook/commits/dirty-quick-fix). However this would brake backward compatibility if merged. I am willing to contribute some time to implement proper fix but for that I would like to have a discussion with a maintainer of the plugin, to ensure we come up with an acceptable solution.

**Comment by jmecosta** [comment]
@auriusbendikas yes i am aware of this limitation, the implemented solution was for our current use case. So yes it would be nice to support any kind of branch definitions. the proper solution is to support any branch definition, if you come up with any good suggestion please share. your quick-fix cannot be merged since it will brake existing func as you mention.

**Comment by auriusbendikas** [comment]
Before I start any work I would like to discuss the changes I am planning to do:

1. Move "Trigger Type" configuration from "Connection" tab, to "Build Configurations".
0. Refactor "Build Configurations" tab to use the same configuration concept as in "External Triggers". The interface will be divided into 3 blocks.
   - Input Fields - "Trigger Type", "Branch Pattern", "Target ID".
   - "Add"/"Remove" buttons.
   - The "Build Configurations" table with "Trigger Type", "Branch Pattern", "Target ID" columns.
0. Explanation of configuration items in "Build Configurations" tab.
    - "Trigger Type" - drop down box with values "build" or "vcs".
    - "Branch Pattern" - input box with RegExp pattern. The pattern depends from the value in "Trigger Type"
      - IF it is "vcs" then it is normal RegExp pattern without any capturing blocks.
      - IF it is "build" then it must contain one capturing block to capture branch name to be sent to TC.
    - "Target ID" - input box with either VCS Roots Ids or Build Configuration Ids depending on the "Trigger Type" value.

This will be quite drastic change and would brake existing configurations. This would require all users to reconfigure their configurations after upgrade. But is is most flexible yet complicated solution (in my opinion).

Some related questions:

1. What does "Dependencies" mean in "TeamCity Triggers" and "External Triggers" tab?
0. What does "Configurations" mean in "TeamCity Triggers" tab?

P.S.: This would also solve #14.

**Comment by jmecosta** [comment]
alright, i like the implementation. I makes sense to make it extendable to any type of branch configuration. 

to your questions. 

1. we internally are running integration tests on demand, these cant be run for every feature branch so we allow users to define a set of dependencies > build configuration so they can trigger the tests manually. This can only be done if the dependency is build since usually the integration test requires a artifact from the dependency. if you set a few then in the built & test menu you will have those available. The external triggers is similar but is meant to sent rest requests to external systems in this case there is no reporting back in the pull request.
2. the configurations are the build configurations that you can remote trigger, as explained in 1.

this tab would also benifict from the same kind of approach you are applying to the remote triggers. but that can be done later also.

**Comment by auriusbendikas** [comment]
Hi,

I have noticed that contributions to this project does not follow same coding guidelines (Indentation, line length etc.). I am using eclipse and always configure source code auto-format on every save I make. I wanted to create code formatting profile, but now I am confused witch style to use ([ConcreteSettingsService.java](/TrimbleSolutionsCorporation/TeamcityTriggerHook/blob/master/src/main/java/com/trimble/tekla/ConcreteSettingsService.java) vs [TeamcityConnector.java](/TrimbleSolutionsCorporation/TeamcityTriggerHook/blob/master/src/main/java/com/trimble/tekla/teamcity/TeamcityConnector.java)).
* Can you shortly describe expected coding style, or point out some files I could use as a reference?
* Can I reformat coding style of existing files I am going to change? Or I should maintain the same style that was originally there?

**Comment by jmecosta** [comment]
the concretesettings has the correct style, perhaps my netbeans misbehave in the connector. we 2 spaces for indentation level. feel free to adjust those, or open a ticket so that it can be fixed later
